FirebaseMessaging/Sources/FIRMessagingContextManagerService.m

```diff
@@ -129,7 +136,70 @@ + (BOOL)handleContextManagerLocalTimeMessage:(NSDictionary *)message {
+  NSCalendar *calendar = [NSCalendar currentCalendar];
+  if (@available(macOS 10.14, iOS 10.0, watchOS 3.0, tvOS 10.0, *)) {
+    NSCalendarUnit unit = NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay |
+                          NSCalendarUnitHour | NSCalendarUnitMinute | NSCalendarUnitSecond;
+    NSDateComponents *dateComponents = [calendar components:(NSCalendarUnit)unit fromDate:date];
+    UNCalendarNotificationTrigger *trigger =
+        [UNCalendarNotificationTrigger triggerWithDateMatchingComponents:dateComponents
+                                                                 repeats:YES];
+
+    UNMutableNotificationContent *content = [[UNMutableNotificationContent alloc] init];
+    NSDictionary *apsDictionary = message;
+
+    // Badge is universal
+    if (apsDictionary[kFIRMessagingContextManagerBadgeKey]) {
+      content.badge = apsDictionary[kFIRMessagingContextManagerBadgeKey];
+    }
+#if TARGET_OS_IOS || TARGET_OS_OSX || TARGET_OS_WATCH
+    // The following fields are not available on tvOS
+    if ([apsDictionary[kFIRMessagingContextManagerBodyKey] length]) {
+      content.body = apsDictionary[kFIRMessagingContextManagerBodyKey];
+    }
+    if ([apsDictionary[kFIRMessagingContextManagerTitleKey] length]) {
+      content.title = apsDictionary[kFIRMessagingContextManagerTitleKey];
+    }
+
+    if (apsDictionary[kFIRMessagingContextManagerSoundKey]) {
+      content.sound = apsDictionary[kFIRMessagingContextManagerSoundKey];
+    }
+
+    if (apsDictionary[kFIRMessagingContextManagerCategoryKey]) {
+      content.categoryIdentifier = apsDictionary[kFIRMessagingContextManagerCategoryKey];
+    }
+
+    NSDictionary *userInfo = [self parseDataFromMessage:message];
+    if (userInfo.count) {
+      content.userInfo = userInfo;
+    }
+#endif
+    NSString *identifier = apsDictionary[kFIRMessagingID];
+    if (!identifier) {
+      identifier = [NSUUID UUID].UUIDString;
+    }
+
+    UNNotificationRequest *request = [UNNotificationRequest requestWithIdentifier:identifier
+                                                                          content:content
+                                                                          trigger:trigger];
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    [center addNotificationRequest:request
+             withCompletionHandler:^(NSError *_Nullable error) {
+               if (error) {
+                 FIRMessagingLoggerError(
+                     kFIRMessagingMessageCodeContextManagerServiceFailedLocalSchedule,
+                     @"Failed scheduling local timezone notification: %@.", error);
+               }
+             }];
+  }
+}
+
 + (void)scheduleLocalNotificationForMessage:(NSDictionary *)message atDate:(NSDate *)date {
+  if (@available(macOS 10.14, iOS 10.0, watchOS 3.0, tvOS 10.0, *)) {
+    [self scheduleiOS10LocalNotificationForMessage:message atDate:date];
```

Comment:
I wonder if we need to make any validation of the date (e.g. if it is in the past or in a distant future)?

Reply:
We do check it here: https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseMessaging/Sources/FIRMessagingContextManagerService.m#L99

And if it's very near future we schedule right away.